tsl/src/remote/dist_copy.c

```diff
@@ -797,36 +1115,230 @@ send_copy_data(StringInfo row_data, const List *connections)
+ */
+typedef struct DataNodeRows
+{
+	int data_node_id;
+	Oid server_oid;
+	TSConnection *connection;
+	int rows_total;
+
+	/* Array of indices into the batch row array. */
+	int *row_indices;
+} DataNodeRows;
+
 static bool
 remote_copy_process_and_send_data(RemoteCopyContext *context)
 {
 	Hypertable *ht = context->ht;
-	Chunk *chunk;
-	Point *point;
-	const List *connections;
+	const int n = context->current_batch_rows;
+	Assert(n <= MAX_BATCH_ROWS);
 
-	if (context->binary_operation)
-		point = get_current_point_for_binary_copy(context->data_context, ht->space);
-	else
-		point = get_current_point_for_text_copy(context->data_context);
+	/*
+	 * This list tracks the per-batch insert states of the data nodes
+	 * (DataNodeRows).
+	 */
+	List *data_nodes = NIL;
 
-	chunk = get_target_chunk(ht, point, &context->connection_state);
-	connections = get_connections_for_chunk(context, chunk->fd.id, chunk->data_nodes, GetUserId());
+	/* For each row, find or create the destination chunk. */
+	bool did_flush = false;
+	for (int row_in_batch = 0; row_in_batch < n; row_in_batch++)
+	{
+		Point *point = context->batch_points[row_in_batch];
 
-	/* for remote copy, we don't use chunk insert states on the AN.
-	 * so we need to explicitly set the chunk as unordered when copies
-	 * are directed to previously compressed chunks
+		Chunk *chunk = ts_hypertable_find_chunk_for_point(ht, point);
+		if (chunk == NULL)
+		{
+			if (!did_flush)
+			{
+				/*
+				 * The data node connections have to be flushed before creating
+				 * a new chunk. They might have outstanding COPY data from the
+				 * previous batch.
+				 */
+				flush_active_connections(&context->connection_state);
+				did_flush = true;
+			}
+			chunk = ts_hypertable_create_chunk_for_point(ht, point);
+		}
+
+		/*
+		 * For remote copy, we don't use chunk insert states on the AN.
+		 * So we need to explicitly set the chunk as unordered when copies
+		 * are directed to previously compressed chunks.
+		 */
+		if (ts_chunk_is_compressed(chunk) && (!ts_chunk_is_unordered(chunk)))
+			ts_chunk_set_unordered(chunk);
+
+		/*
+		 * Schedule the row for sending to the data nodes containing the chunk.
+		 */
+		ListCell *lc;
+		foreach (lc, chunk->data_nodes)
+		{
+			ChunkDataNode *chunk_data_node = lfirst(lc);
+			/* Find the existing insert state for this data node. */
+			DataNodeRows *data_node_rows = NULL;
+			ListCell *lc2;
+			foreach (lc2, data_nodes)
+			{
+				data_node_rows = lfirst(lc2);
+				if (chunk_data_node->foreign_server_oid == data_node_rows->server_oid)
+				{
+					break;
+				}
+			}
+
+			if (lc2 == NULL)
+			{
+				/* No insert state for this data node yet. Create it. */
+				data_node_rows = palloc(sizeof(DataNodeRows));
+				data_node_rows->server_oid = chunk_data_node->foreign_server_oid;
+				data_node_rows->rows_total = 0;
+				data_node_rows->row_indices = palloc(sizeof(int) * context->current_batch_rows);
+				data_nodes = lappend(data_nodes, data_node_rows);
+			}
```

Comment:
It's a matter of taste, but to me a wall of code is more readable than a sequence of function calls. If it were used in several places, I'd extract it, but the entire DataNodeRows stuff is private to this function. Postgres follows the same approach -- not extracting functions just because the parent function is too long.

Reply:
Mostly concerned about code duplication.

For shorter swaths of code I do not think moving it to a function makes a difference, for anything that you can't immediately grasp with a single look, or which requires a lot of scrolling, I usually factor it out.